src/core/Akka.Streams/Dsl/Valve.cs

```diff
@@ -0,0 +1,181 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="Valve.cs" company="Akka.NET Project">
+//     Copyright (C) 2009-2018 Lightbend Inc. <http://www.lightbend.com>
+//     Copyright (C) 2013-2018 .NET Foundation <https://github.com/akkadotnet/akka.net>
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Threading.Tasks;
+using Akka.Streams.Stage;
+
+namespace Akka.Streams.Dsl
+{
+    /// <summary>
+    /// Mode for <see cref="IValveSwitch"/>
+    /// </summary>
+    public enum SwitchMode
+    {
+        Open,
+        Close
+    }
+
+    /// <summary>
+    /// Pause/resume a Flow
+    /// </summary>
+    public interface IValveSwitch
+    {
+        /// <summary>
+        /// Change the state of the valve
+        /// </summary>
+        /// <param name="mode">expected mode to switch on</param>
+        /// <returns>A task that completes with true if the mode did change and false if it already was in the requested mode</returns>
+        Task<bool> Flip(SwitchMode mode);
+
+        /// <summary>
+        /// Obtain the state of the valve
+        /// </summary>
+        /// <returns>A task that completes with <see cref="SwitchMode"/> to indicate the current state of the valve</returns>
+        Task<SwitchMode> GetMode();
+    }
+
+    class ValveSwitch : IValveSwitch
+    {
+        private readonly Action<Tuple<SwitchMode, TaskCompletionSource<bool>>> _flipCallback;
+        private readonly Action<TaskCompletionSource<SwitchMode>> _getModeCallback;
+
+        public ValveSwitch(Action<Tuple<SwitchMode, TaskCompletionSource<bool>>> flipCallback, Action<TaskCompletionSource<SwitchMode>> getModeCallback)
+        {
+            _flipCallback = flipCallback;
+            _getModeCallback = getModeCallback;
+        }
+
+        public Task<bool> Flip(SwitchMode flipToMode)
+        {
+            var promise = new TaskCompletionSource<bool>();
+            _flipCallback(Tuple.Create(flipToMode, promise));
+            return promise.Task;
```

Comment:
either use `promise` or `completion` as name in both cases, I would prefer `completion` ;-)